Lazily evaluate NonNullTypes

jcouv opened the pull request:
This PR reverts the cycle mitigations put in place in https://github.com/dotnet/roslyn/pull/28320 and delays the evaluation of `NonNullTypes` by using `INonNullTypesContext`. 
This generally holds the symbol who can give us the answer, but in some cases, we also use some dummy contexts (which always answer `NonNullTypes==true` or `NonNullTypes==false`).

I'm still working on a couple of test regressions (in OHI).

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/SemanticFacts.cs:
```diff
@@ -8,7 +8,8 @@
 {
-    internal partial class Symbol
+
+    internal partial class Symbol : INonNullTypesContext
```
> : INonNullTypesContext [](start = 33, length = 23)

Consider moving to Symbol.cs.

cston commented on src/Compilers/CSharp/Portable/Symbols/Symbol.cs:
```diff
@@ -837,7 +837,7 @@ public virtual bool HasUnsupportedMetadata
         /// Not valid to call on non-definitions.
         /// </summary>
-        internal virtual bool NonNullTypes
+        public virtual bool NonNullTypes
```
>public [](start = 8, length = 6)

Why `public`?

jcouv replied:
It's `public` to implicitly implement the `INonNullTypesContext` interface.

cston commented on src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs:
```diff
@@ -79,14 +79,9 @@ public sealed override int GetHashCode()
             compilerInternalOptions: SymbolDisplayCompilerInternalOptions.IncludeNonNullableTypeModifier);
 
-        internal static TypeSymbolWithAnnotations CreateNonNull(bool nonNullTypes, TypeSymbol typeSymbol)
+        internal static TypeSymbolWithAnnotations CreateUnannotated(INonNullTypesContext nonNullTypesContext, TypeSymbol typeSymbol)
```
>CreateUnannotated [](start = 50, length = 17)

Thanks for renaming.

cston commented on src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs:
```diff
@@ -107,21 +102,25 @@ public static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, ImmutableA
             }
 
-            return new NonLazyType(typeSymbol, isNullable: typeSymbol.IsNullableType(), customModifiers);
+            return Create(typeSymbol, typeSymbol.IsNullableType());
```
>Create [](start = 19, length = 6)

We're dropping `customModifiers`.

cston commented:
>         private static bool IsNullableToIsAnnotated(bool? isNullable) => isNullable == true;

Please remove these two methods. (The first is no longer called, and the second has only one caller.)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs:578 in a8184b4. [](commit_id = a8184b41baa4579af173d8062fb23e663d9c1c62, deletion_comment = False)

cston commented on src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs:
```diff
@@ -208,10 +208,9 @@ public TypeSymbolWithAnnotations AsNullableReferenceOrValueType(CSharpCompilatio
+        /// Lazy [NonNullTypes] context used for determining whether unannotated types are not nullable.
         /// </summary>
-        public abstract bool NonNullTypes { get; }
+        public INonNullTypesContext NonNullTypesContext { get; protected set; }
```
>public INonNullTypesContext NonNullTypesContext { get; protected set; } [](start = 8, length = 71)

We only need a backing field for one of the two derived classes. Consider leaving this `abstract`.

cston commented on src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs:
```diff
@@ -208,10 +208,9 @@ public TypeSymbolWithAnnotations AsNullableReferenceOrValueType(CSharpCompilatio
         /// <summary>
-        /// [NonNullTypes] context used for determining
-        /// whether unannotated types are not nullable.
+        /// Lazy [NonNullTypes] context used for determining whether unannotated types are not nullable.
```
>Lazy [](start = 12, length = 4)

Minor point: Consider removing `Lazy` since the context itself is not lazy, just the calls to the `NonNullTypes` property are.

cston commented on src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs:
```diff
@@ -801,7 +796,7 @@ protected override TypeSymbolWithAnnotations DoUpdate(TypeSymbol typeSymbol, Imm
                 }
 
-                return new NonLazyType(typeSymbol, isNullable: true, customModifiers);
+                return new NonLazyType(typeSymbol, NonNullTypesContext, isAnnotated: IsAnnotated, customModifiers);
```
>NonNullTypesContext [](start = 51, length = 19)

NonNullTypesContextTrue.Instance

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Semantics/INonNullTypesContext.cs:
```diff
@@ -0,0 +1,25 @@
+        bool NonNullTypes { get; }
+    }
+
+    internal sealed class NonNullTypesContextTrue : INonNullTypesContext
```
Nit: Rather than True / False suffix consider `EnableNonNullTypes` / `DisableNonNullTypes`.

jcouv replied:
I was actually thinking to change to `NonNullTypesTrueContext`, since that's what we tend to speak out loud ("[NonNullTypes(true)] context"). We don't typically say "enabled" or "disabled", although it's correct.
Sounds alright?

jaredpar replied:
That sounds fine to me.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Semantics/INonNullTypesContext.cs:
```diff
@@ -0,0 +1,25 @@
+
+    internal sealed class NonNullTypesContextTrue : INonNullTypesContext
+    {
+        public static INonNullTypesContext Instance = new NonNullTypesContextTrue();
```
readonly

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Semantics/INonNullTypesContext.cs:
```diff
@@ -0,0 +1,25 @@
+
+    internal sealed class NonNullTypesContextFalse : INonNullTypesContext
+    {
+        public static INonNullTypesContext Instance = new NonNullTypesContextFalse();
```
readonly

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:
```diff
@@ -222,12 +222,12 @@ internal ImmutableArray<TypeSymbolWithAnnotations> GetTypeParametersAsTypeArgume
             // Resolving [NonNullTypes] only makes sense within the definition of the generic type
             // or method. If this is a substituted symbol, we use the default NonNullTypes context.
```
Is "default NonNullTypes context" accurate here now? It seems like we're using the false context. Or am I just misunderstanding the terminology?

jcouv replied:
"default NonNullTypes context" seems correct, since `[NonNullTypes(false)]` is the root context (outside of the scope of an explicit `[NonNullTypes...]`).
But here it really doesn't matter which context we use for that case. I'll change to say "dummy context".

jcouv replied:
This made me think we shouldn't even use the `NonNullTypesFalseContext`. I created one more fixed context that always throws.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs:
```diff
@@ -113,7 +113,7 @@ private TupleTypeDecoder(ImmutableArray<string> elementNames)
             return (object)decoded == (object)type ?
                 metadataType :
-                TypeSymbolWithAnnotations.Create(decoded, nonNullTypes: metadataType.NonNullTypes, isAnnotated: metadataType.IsAnnotated, metadataType.CustomModifiers);
+                TypeSymbolWithAnnotations.Create(decoded, metadataType.NonNullTypesContext, isAnnotated: metadataType.IsAnnotated, metadataType.CustomModifiers);
```
Consider keeping the named argument style for nonNullTypesContext

jcouv replied:
Whenever the argument is `this` or `type`, I added the named argument. Whenever the argument is `x.NonNullTypesContext`, I felt the named argument was redundant. 
Also, since we're using a specific type (`INonNullTypesContext`) rather than something like a `bool` or `int`, there is little ambiguity or chance of passing to as wrong parameter.

jaredpar replied:
That makes sense.

jcouv commented:
@cston Let me know if you'll want to take a look at the changes since your review. I adjusted a couple of tests that already had PROTOTYPE markers and removed some uses of NonNullTypes.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -4716,9 +4723,10 @@ class B : A
             // PROTOTYPE(NullableReferenceTypes): Should report return type mismatch
             // for M1 and M2 (see https://github.com/dotnet/roslyn/issues/28684).
```
>https://github.com/dotnet/roslyn/issues/28684 [](start = 34, length = 45)

It looks like you've fixed #28684. Is the issue with `M2<S>` a case where the meaning of `S?` is unclear since `S` is unconstrained? If so, that should be resolved when we report an error for use of `T?` for unconstrained `T`.

cston commented on src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:
```diff
@@ -9085,7 +9085,11 @@ End Module
+        ' The NonNullTypes context for nested tuple types is using a dummy rather than actual context from surrounding code.
+        ' This does not affect `IsNullable`, but it affects `IsAnnotatedWithNonNullTypesContext`, which is used in comparisons.
+        ' So when we copy modifiers (re-applying nullability information, including actual NonNullTypes context), we make the comparison fail.
+        ' I think the solution is to never use a dummy context, even for value types.
```
>I think the solution is to never use a dummy context, even for value types. [](start = 10, length = 75)

Consider logging a bug and referencing this test. This seems like an issue we should address soon. Thanks.

jcouv replied:
Agree.
Filed https://github.com/dotnet/roslyn/issues/28766

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/CustomModifierUtils.cs:
```diff
@@ -53,19 +53,19 @@ internal static class CustomModifierUtils
         /// <param name="sourceType">Type that already has custom modifiers.</param>
         /// <param name="destinationType">Same as <paramref name="sourceType"/>, but without custom modifiers.  May differ in object/dynamic.</param>
         /// <param name="containingAssembly">The assembly containing the signature referring to the destination type.</param>
-        /// <param name="destinationUsesNonNullTypes">If the destination is in a context with NonNullTypes set. If true, unannotated references types are treated as non-nullable, otherwise as null-oblivious.</param>
```
><param name="destinationUsesNonNullTypes"> [](start = 12, length = 42)

Does the missing `<param/>` not result in a warning?
